nemo_skills/inference/prompt/openai_chat.yaml

```diff
@@ -0,0 +1,7 @@
+delimiter: "\n\n\n\n\n\n"
```

Comment:
So it looks like this format is still not general enough to support all of the use-cases we have. Here are the things
that we need to be able to represent:

1. No "chat" format

2. NeMo chat format where explicit tokens need to be specified

3. Non-nemo chat models, where system/user might have different separator format, e.g.
   ```
       "<<SYS>>\n{system}\n<</SYS>>\n\n"
       "<|system|>\n{system}</s>\n"
       "[INST] {text} [/INST]",
       "<|user|>\n{text}\n<|assistant|>\n",
  
   ```

3. OpenAI (or other cloud providers) where you don't need to list tokens, but need the message separated into system/user

And in all cases we want a convenient support for few-shot examples, so that prompt can be constructed from them easily.

Here is what I propose to use and configs that are equivalent to currently supported use-cases. Might be necessary
to make some adjustments.

Current code_base.yaml:

```
# template for few-shot examples. 
# We need to separate the last part (actual question that's missing a generated_solution), because
# it needs to follow-up with special tokens for chat models that we have to explicitly handle
few_shot_examples:
  template: "Question:\n{question}\n\n{context}My solution:\n{generated_solution}"
  delimiter: "\n\n\n\n\n\n"
# will generate {examples} field based on the above
# examples = template.format(example1_info) + delimiter + ... + template.format(exampleN_info)

# system part of cloud LLM api call
system: ""

# user part of cloud LLM api call
user: "\
    Here are some examples of questions and solutions followed by a new question that you need to solve.\n\
    Make sure to put the answer (and only answer) inside \\boxed{}.\n\n\
    {examples}\n\n\n\n\n\n\
    Question:\n{question}\n\n{context}"

# ignored for cloud LLMs, but used for NeMo or other local models to explicitly format turn separators
prompt: {system}{user}My solution:\n
```

Current code_sfted.yaml. Note that few-shot examples is not used here, since we never want to show few-shots
for finetuned models (it does not make it any better, but takes up the context)

```
# system part of cloud LLM api call
system: "You're an expert Python programmer and mathematician. \
  Help the user to solve this problem using code when necessary. \
  Make sure to put the answer (and only answer) inside \\boxed{}."

# user part of cloud LLM api call
user: "{question}\n\n{context}"

# ignored for cloud LLMs, but used for NeMo or other local models to explicitly format turn separators
prompt: System:\n{system}\n\nUser:\n{user}Assistant:\n
```

From igitman/chat-format branch code_base.yaml

```
# template for few-shot examples. 
# We need to separate the last part (actual question that's missing a generated_solution), because
# it needs to follow-up with special tokens for chat models that we have to explicitly handle
few_shot_examples:
  template: "Question:\n{question}\n\n{context}My solution:\n{generated_solution}"
  delimiter: "\n\n\n\n\n\n"
# will generate {examples} field based on the above
# examples = template.format(example1_info) + delimiter + ... + template.format(exampleN_info)

# system part of cloud LLM api call
system: ""

# user part of cloud LLM api call
user: "\
    Here are some examples of questions and solutions followed by a new question that you need to solve.\n\
    Make sure to put the answer (and only answer) inside \\boxed{}.\n\n\
    {examples}\n\n\n\n\n\n\
    Question:\n{question}\n\n{context}"

# ignored for cloud LLMs, but used for NeMo or other local models to explicitly format turn separators
prompt: <extra_id_0>System\n{system}\n<extra_id_1>User\n{user}<extra_id_1>Assistant\nMy solution:\n
```